File: src/handler/catalogue.cpp
#include "catalogue.h"

#include "catalogue/basictype.h"
#include "core/application.h"
#include "catalogue/catalogue.h"

#include "queryprocessing/ddlsemantic.h"

namespace AdaptiveDB
{
    void getAllTables(Request &req, nlohmann::json &res) {
        auto catalogue = Application::instance().catalogue();
        res["tables"] = nlohmann::json::array();
        for (const auto &table : catalogue->tables()) {
            nlohmann::json tableJson;
            tableJson["name"] = table.name;
            nlohmann::json columnsJson;
            for (const auto &column : table.columns) {
                nlohmann::json columnJson;
                columnJson["name"] = column.name;
                columnJson["type"] = basicTypeStrings[column.type];
                columnJson["nullable"] = column.nullable;
                if (table.primaryKey == &column - &table.columns[0]) {
                    columnJson["primary"] = true;
                }
                columnsJson.push_back(columnJson);
            }
            tableJson["columns"] = columnsJson;
            res["tables"].push_back(tableJson);
        }
    }

    void createTableBySchema(Request &req, nlohmann::json &res)
    {
        auto body = req.body;
        if (!body.contains("schema"))
        {
            res["status"] = "ERROR";
            res["message"] = "Body requires a 'schema' field";
            return;
        }

        DDLLexer lexer(body["schema"]);
        auto tokens = lexer.tokenize();

        if (lexer.errors().size() > 0)
        {
            nlohmann::json errors;
            for (auto &error : lexer.errors())
            {
                errors.push_back(error);
            }
            res["status"] = "ERROR";
            res["message"] = "Lexer error";
            res["errors"] = errors;
            return;
        }

        lexer = DDLLexer(body["schema"]);
        DDLParser parser(lexer);
        auto models = parser.parseModels();
        
        if (parser.errors().size() > 0)
        {
            nlohmann::json errors;
            for (auto &error : parser.errors())
            {
                errors.push_back(error);
            }
            res["status"] = "ERROR";
            res["message"] = "Parser error";
            res["errors"] = errors;
            return;
        }

        DDLSemanticChecker checker;
        checker.checkModels(models);

        if (checker.errors().size() > 0)
        {
            nlohmann::json errors;
            for (auto &error : checker.errors())
            {
                errors.push_back(error);
            }
            res["status"] = "ERROR";
            res["message"] = "Semantic error";
            res["errors"] = errors;
            return;
        }

        for (const auto &model : models)
        {
            CatalogueTable table;
            table.name = model.name;
            int i = 0;
            for (const auto &field : model.fields)
            {
                CatalogueColumn tableColumn;
                tableColumn.name = field.name;
                tableColumn.type = field.type;
                tableColumn.nullable = field.nullable;
                if (field.primary)
                {
                    table.primaryKey = i;
                }
                table.columns.push_back(tableColumn);
                i++;
            }
            Application::instance().catalogue()->addTable(table);
        }


        res["status"] = "OK";
        res["message"] = "Tables created";
    }
}

File: src/handler/catalogue.rs
use http_body_util::{combinators::BoxBody, BodyExt};
use hyper::{body::Bytes, Method, Request, Response, StatusCode};

use crate::{catalogue, empty, full, queryprocessing};

pub async fn catalogue_handler(
    req: Request<hyper::body::Incoming>,
) -> Result<Response<BoxBody<Bytes, hyper::Error>>, hyper::Error> {
    match (req.method(), req.uri().path()) {
        (&Method::POST, "/catalogue") => post_catalogue(req).await,
        // Return 404 Not Found for other routes.
        _ => {
            let mut not_found = Response::new(empty());
            *not_found.status_mut() = StatusCode::NOT_FOUND;
            Ok(not_found)
        }
    }
}

#[derive(Debug, serde::Deserialize)]
struct PostTable {
    schema: String,
}

#[derive(Debug, serde::Serialize)]
struct PostTableErrorResponse {
    errors: Vec<String>,
}

async fn post_catalogue(
    req: Request<hyper::body::Incoming>,
) -> Result<Response<BoxBody<Bytes, hyper::Error>>, hyper::Error> {
    let body = req.into_body().collect().await?.to_bytes();
    let body_json = serde_json::from_slice::<PostTable>(&body);
    if body_json.is_err() {
        return Ok(Response::builder()
            .status(StatusCode::BAD_REQUEST)
            .body(full(body_json.as_ref().unwrap_err().to_string()))
            .unwrap());
    }

    let body_json = body_json.unwrap();
    let schema = body_json.schema;

    let models = queryprocessing::ddl::parse(schema);
    let errors = queryprocessing::ddl::analyze(&models);
    if !errors.is_empty() {
        return Ok(Response::builder()
            .status(StatusCode::BAD_REQUEST)
            .body(full(
                serde_json::to_string(&PostTableErrorResponse { errors }).unwrap(),
            ))
            .unwrap());
    }

    let mut catalogue = crate::get_catalogue().lock().await;
    for model in models.iter() {
        if catalogue.table_exists(&model.name) {
            return Ok(Response::builder()
                .status(StatusCode::BAD_REQUEST)
                .body(full(format!(
                    "Aborted. Table {} already exists",
                    model.name
                )))
                .unwrap());
        }
    }

    for model in models.iter() {
        let columns = model
            .fields
            .iter()
            .map(|field| {
                catalogue::Column::new(
                    field.name.clone(),
                    field.field_type.clone(),
                    field.is_nullable,
                )
            })
            .collect();
        let primary_key_index = model
            .fields
            .iter()
            .position(|field| field.is_primary_key)
            .expect("Analyzing should have caught this error");
        let table = catalogue::Table::new(model.name.clone(), columns, primary_key_index as u32);
        catalogue.add_table(table);
    }

    catalogue.save().unwrap();

    let model_names: Vec<String> = models.iter().map(|model| model.name.clone()).collect();
    Ok(Response::builder()
        .status(StatusCode::CREATED)
        .body(full(format!("Created following tables: {:?}", model_names)))
        .unwrap())
}
